### PR TITLE
Autoconfigurable Telemetry SDK support

### PR DESCRIPTION
**Pull request by Shephalimittal**
### Description
Added support for OpenTelemetry SDK Autoconfigure. 
This artifact implements environment-based autoconfiguration of the OpenTelemetry SDK. This can be an alternative to programmatic configuration using the normal SDK builders.

All options support being passed as Java system properties, e.g., -Dotel.traces.exporter=zipkin or environment variables, e.g., OTEL_TRACES_EXPORTER=zipkin. For more details refer [here](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/README.md).

The intention here is to remove managing tracing configuration like exporter, endpoints, processor delays, sampler etc through Opensearch cluster settings. Also, if there are any custom exporter or sampler settings, user can pass that through jvm options or environment variables.

This also provides support to add custom exporters. InMemorySpanExporter is the exporter used for integration tests, have made it autoconfigurable by using ConfigurableSpanExporterProvider class

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: distribution/src/bin/opensearch-service-mgr.exe
[the remark this answers is not in the document]
Comment (Shephalimittal) [reply distribution/src/bin/opensearch-service-mgr.exe:1]: Ack, Removed these.

File: plugins/telemetry-otel/build.gradle [hunk 1]
```diff
@@ -79,3 +75,20 @@ tasks.register("writeTestJavaPolicy") {
...
 }
+
+tasks.named("internalClusterTest").configure {
+  systemProperty 'otel.traces.exporter', 'InMemorySpanExporter'
+  final File tmp = file("${buildDir}/tmp")
+  if (tmp.exists() == false && tmp.mkdirs() == false) {
+    throw new GradleException("failed to create temporary directory [${tmp}]")
+  }
+  final File javaPolicy = file("${tmp}/java.policy")
+  javaPolicy.write(
```
[the remark this answers is not in the document]
Comment (Shephalimittal) [reply plugins/telemetry-otel/build.gradle:86]: yes. These permissions are added for integ tests

File: plugins/telemetry-otel/src/main/java/org/opensearch/telemetry/tracing/OTelResourceProvider.java [hunk 2]
```diff
@@ -8,72 +8,32 @@
...
 public final class OTelResourceProvider {
     private OTelResourceProvider() {}
 
     /**
-     * Creates OpenTelemetry instance with default configuration
-     * @param settings cluster settings
+     * Creates OpenTelemetry instance with AutoConfiguredOpenTelemetrySdk settings configuration
      * @return OpenTelemetry instance
      */
-    public static OpenTelemetry get(Settings settings) {
```
[the remark this answers is not in the document]
Comment (Shephalimittal) [reply plugins/telemetry-otel/src/main/java/org/opensearch/telemetry/tracing/OTelResourceProvider.java:43]: Ack. Added cluster settings back.

File: plugins/telemetry-otel/src/main/java/org/opensearch/telemetry/tracing/OTelResourceProvider.java [hunk 3]
```diff
@@ -8,72 +8,32 @@
...
-            .addSpanProcessor(spanProcessor(settings, spanExporter))
-            .setResource(resource)
-            .setSampler(sampler)
-            .build();
-
-        return OpenTelemetrySdk.builder().setTracerProvider(sdkTracerProvider).setPropagators(contextPropagators).buildAndRegisterGlobal();
+    public static OpenTelemetry get() {
+        OpenTelemetry openTelemetry = AutoConfiguredOpenTelemetrySdk.builder()
+            .setResultAsGlobal(false)
+            .addPropertiesSupplier(() -> Map.of("otel.logs.exporter", "none", "otel.metrics.exporter", "none"))
```
[the remark this answers is not in the document]
Comment (Shephalimittal) [reply plugins/telemetry-otel/src/main/java/org/opensearch/telemetry/tracing/OTelResourceProvider.java:34]: Adding ```otel.logs.exporter``` and ```otel.metrics.exporter``` as none. ```otel.traces.exporter``` can be added in jvm options.

**Comment by Shephalimittal** [comment]
@reta Please review.

File: plugins/telemetry-otel/src/main/java/org/opensearch/telemetry/tracing/OTelResourceProvider.java [hunk 4]
```diff
@@ -9,71 +9,34 @@
...
-    public static OpenTelemetry get(Settings settings, SpanExporter spanExporter, ContextPropagators contextPropagators, Sampler sampler) {
-        Resource resource = Resource.create(Attributes.of(ResourceAttributes.SERVICE_NAME, "OpenSearch"));
-        SdkTracerProvider sdkTracerProvider = SdkTracerProvider.builder()
-            .addSpanProcessor(spanProcessor(settings, spanExporter))
-            .setResource(resource)
-            .setSampler(sampler)
-            .build();
-
-        return OpenTelemetrySdk.builder().setTracerProvider(sdkTracerProvider).setPropagators(contextPropagators).buildAndRegisterGlobal();
+        OpenTelemetry openTelemetry = AutoConfiguredOpenTelemetrySdk.builder()
```
Comment (reta) [inline plugins/telemetry-otel/src/main/java/org/opensearch/telemetry/tracing/OTelResourceProvider.java:32]: That will solve your issues with security manager:

```
return AccessController.doPrivileged(
            (PrivilegedAction<OpenTelemetry>) () -> AutoConfiguredOpenTelemetrySdk.builder()
                .setResultAsGlobal(false)
                .addPropertiesSupplier(() -> Map.of("otel.logs.exporter", "none", "otel.metrics.exporter", "none"))
                .addResourceCustomizer(
                    (resource, config) -> resource.merge(Resource.builder().put(ResourceAttributes.SERVICE_NAME, "OpenSearch").build())
                )
                .build()
                .getOpenTelemetrySdk()
        );
```

**Comment by reta** [comment]
@Shephalimittal OTel auto configuration is certainly useful for application developers, now I am struggling to understand how that will work in OpenSeach:
 - user installs OpenSeach
 - user install OpenSearch Otel plugin

User wants to configure OpenSearch Otel plugin to send traces to Jaeger (for example), how that could be accomplished with auto configuration?

**Comment by Gaganjuneja** [comment]
> @Shephalimittal OTel auto configuration is certainly useful for application developers, now I am struggling to understand how that will work in OpenSeach:
> 
> * user installs OpenSeach
> * user install OpenSearch Otel plugin
> 
> User wants to configure OpenSearch Otel plugin to send traces to Jaeger (for example), how that could be accomplished with auto configuration?

As per my understanding, user will have to provide the jaeger related configurations in the jvm.options file.

**Comment by reta** [comment]
> As per my understanding, user will have to provide the jaeger related configurations in the jvm.options file.

If that is the case (which could only be done for some properties), this is certainly no go: we have established settings management.

**Comment by Gaganjuneja** [comment]
> > As per my understanding, user will have to provide the jaeger related configurations in the jvm.options file.
> 
> If that is the case (which could only be done for some properties), this is certainly no go: we have established settings management.

Only challenge with doing setting management ourselves is we always need to sync with the OTEL. The OTLPGrpcSpanExporter itself has more than 10 settings and similarly BatchSpanProcessor. So we need to expose these all through OpenSearch settings and if there are new types of exporters we might have to add support.

OTelAutoConfigureSDK reads the settings/configuration from System properties so not sure if opensearch.yml can also be used for the same.

**Comment by reta** [comment]
> @reta Please review.

@Shephalimittal please answer my questions (https://github.com/opensearch-project/OpenSearch/pull/9453#issuecomment-1690182246)

**Comment by Shephalimittal** [comment]
> @Shephalimittal OTel auto configuration is certainly useful for application developers, now I am struggling to understand how that will work in OpenSeach:
> 
> * user installs OpenSeach
> * user install OpenSearch Otel plugin
> 
> User wants to configure OpenSearch Otel plugin to send traces to Jaeger (for example), how that could be accomplished with auto configuration?

Yes, that will be achieved through this setting ```-Dotel.traces.exporter=jaeger``` in jvm.options. For details refer [link](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/README.md)

Also, if user wants to set up a custom exporter, that can be provided using implementing ConfigurableSpanExporterProvider. I have added InMemorySingletonSpanExporterProvider for integration test.

**Comment by reta** [comment]
Thanks @Shephalimittal 

> Yes, that will be achieved through this setting `-Dotel.traces.exporter=jaeger` in jvm.options. For details refer [link](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/README.md)

We use jvm.options to configure the JVM parameters for JVM process, not plugins / module or extensions: we use  `opensearch.yaml` for configuration of deployment specific setting and this is what the `otel-plugin` should also do. 

> Also, if user wants to set up a custom exporter, that can be provided using implementing ConfigurableSpanExporterProvider. I have added InMemorySingletonSpanExporterProvider for integration test.

Very well, but in most cases the additional dependency (primarily, exporters) have to be provided (and we cannot bundle everything), so it brings us to the the following question: how user should do that? Manually copying JAR files does not look like a viable option at all (since those may require additional dependencies).

**Comment by Shephalimittal** [comment]
> > Only challenge with doing setting management ourselves is we always need to sync with the OTEL. The OTLPGrpcSpanExporter itself has more than 10 settings and similarly BatchSpanProcessor. So we need to expose these all through OpenSearch settings and if there are new types of exporters we might have to add support.
> 
> > Thanks @Shephalimittal
> > > Yes, that will be achieved through this setting `-Dotel.traces.exporter=jaeger` in jvm.options. For details refer [link](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/README.md)
> > 
> > 
> > We use jvm.options to configure the JVM parameters for JVM process, not plugins / module or extensions: we use `opensearch.yaml` for configuration of deployment specific setting and this is what the `otel-plugin` should also do.
> > But problem with using cluster settings is : (Pasting Gagan's point from above)
> > Only challenge with doing setting management ourselves is we always need to sync with the OTEL. The OTLPGrpcSpanExporter itself has more than 10 settings and similarly BatchSpanProcessor. So we need to expose these all through OpenSearch settings and if there are new types of exporters we might have to add support.
> 
> > > Also, if user wants to set up a custom exporter, that can be provided using implementing ConfigurableSpanExporterProvider. I have added InMemorySingletonSpanExporterProvider for integration test.
> > 
> > 
> > Very well, but in most cases the additional dependency (primarily, exporters) have to be provided (and we cannot bundle everything), so it brings us to the the following question: how user should do that? Manually copying JAR files does not look like a viable option at all (since those may require additional dependencies).

Exploring the option to use AffixSettings here for telemetry settings regarding exporters, delays, timeouts, samplers etc and passing these settings as configProperties to AutoConfiguredOpenTelemetrySdk. @reta Any Thoughts?

**Comment by reta** [comment]
> Exploring the option to use AffixSettings here for telemetry settings regarding exporters, delays, timeouts, samplers etc and passing these settings as configProperties to AutoConfiguredOpenTelemetrySdk. @reta Any Thoughts?

@Shephalimittal we need more than that:
 - dependencies (new exporters and samplers)
 - security policy changes

In general, I think we have to take a step back and open RFC issue to outline the vision regarding extensibiltity / customization of the out OTEL plugin (I personally don't see the large picture).

**Comment by ticheng-aws** [comment]
Hi @Shephalimittal, the PR is stalled. Is this being worked upon? Feel free to reach out to maintainers for further reviews.

**Comment by kotwanikunal** [comment]
@Shephalimittal Closing this stalled PR. Please reopen to continue with the contribution.

**Pull request closed without merging**